oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileStoreSplitter.java

```diff
@@ -0,0 +1,291 @@
+import java.nio.charset.Charset;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.apache.commons.io.FileUtils;
+import org.apache.jackrabbit.oak.api.CommitFailedException;
+import org.apache.jackrabbit.oak.api.PropertyState;
+import org.apache.jackrabbit.oak.api.Type;
+import org.apache.jackrabbit.oak.index.IndexHelper;
+import org.apache.jackrabbit.oak.index.IndexerSupport;
+import org.apache.jackrabbit.oak.index.indexer.document.NodeStateEntry;
+import org.apache.jackrabbit.oak.plugins.index.search.Aggregate;
+import org.apache.jackrabbit.oak.plugins.index.search.IndexDefinition;
+import org.apache.jackrabbit.oak.plugins.memory.MemoryNodeStore;
+import org.apache.jackrabbit.oak.query.NodeStateNodeTypeInfoProvider;
+import org.apache.jackrabbit.oak.query.ast.NodeTypeInfo;
+import org.apache.jackrabbit.oak.query.ast.NodeTypeInfoProvider;
+import org.apache.jackrabbit.oak.spi.state.NodeBuilder;
+import org.apache.jackrabbit.oak.spi.state.NodeState;
+import org.apache.jackrabbit.oak.spi.state.NodeStore;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class FlatFileStoreSplitter {
+    private static final Logger log = LoggerFactory.getLogger(FlatFileStoreSplitter.class);
+    private final File workDir;
+    private final boolean useZip = Boolean.parseBoolean(System.getProperty(OAK_INDEXER_USE_ZIP, "true"));
+    private final int splitSize = Integer.getInteger(PROP_SPLIT_STORE_SIZE, DEFAULT_NUMBER_OF_SPLIT_STORE_SIZE);
+    private final IndexerSupport indexerSupport;
+    private final IndexHelper indexHelper;
+    private final Charset charset = UTF_8;
+    private final NodeTypeInfoProvider infoProvider;
+    private final FlatFileStore ffs;
+    private final NodeStore store;
+    public final String splitDirName = "split";
+    private final NodeStateEntryReader entryReader;
+    private long minimumSplitThreshold = 10 * FileUtils.ONE_MB;
+
+    public FlatFileStoreSplitter(FlatFileStore ffs, IndexHelper indexHelper, IndexerSupport indexerSupport) {
+        this.ffs = ffs;
+        this.indexerSupport = indexerSupport;
+        this.indexHelper = indexHelper;
+        this.workDir = new File(indexHelper.getWorkDir(), splitDirName);
+
+        this.store = new MemoryNodeStore(indexerSupport.retrieveNodeStateForCheckpoint());
+        this.infoProvider = new NodeStateNodeTypeInfoProvider(store.getRoot());
+        this.entryReader = new NodeStateEntryReader(indexHelper.getGCBlobStore());
+    }
+
+    public FlatFileStoreSplitter(FlatFileStore ffs, IndexHelper indexHelper, IndexerSupport indexerSupport, long minimumSplitThreshold) {
+        this(ffs, indexHelper, indexerSupport);
+        this.minimumSplitThreshold = minimumSplitThreshold;
+    }
+
+    public Set<IndexDefinition> getIndexDefinitions() throws IOException, CommitFailedException {
```

Comment:
@Ewocker except the split, for others methods not be used outside, it'd better making private in the beginning

Reply:
updated